### ISSUE TITLE
PID generation rules?

### ISSUE THREAD

**Issue by netsensei** [PACKED-vzw/resolver#38]
I'm trying to understand how the resolver is transforming an identification number into a valid PID. Or better: what rules the resolver is following to perform the transformation.

I decided to look at the resolver code directly, but lacking enough inline documentation, several parts are a bit of a mystery to me.

This function in resolver/util.py is responsible for the transformation:

``` python
_clean_re = re.compile(r'[\t !"#$%&\'()*/<=>?@\[\\\]^`{|}]+')
def cleanID(ID):
    patterns = [
        # Exceptions
        ('- ','-'),(' -','-'),('\)+$',''),('\]+$', ''),('\°+$', ''),
        # Simple replacements
        ('\.','_'),(' ','_'),('\(','_'),('\)','_'),('\[','_'),('\]','_'),
        ('\/','_'),('\?','_'),(',','_'),('&','_'),('\+','_'),('°','_'),
        # Replace 1 or more underscores by a single underscore
        ('_+', '_')]
    partial = reduce(lambda str, t: re.sub(t[0], t[1], str),
                     patterns,
                     ID)
    # For safety, let's give it another scrub.
    result = []
    for word in _clean_re.split(partial):
        result.extend(unidecode(word).split())

    return unicode(''.join(result))
```

The first part is pretty much clear. The ID string is run through a reduce function. On each iteration, a pattern matching function is applied and any matches are replaced with an underscore.

Sidenote: I was wondering why you don't do just this, instead of using a `reduce` map function. I assume the latter approach iterates and applies `sub` for each pattern in patterns which seems far more expensive then just applying `sub` twice.

``` python
partial = re.sub([\]+\)+\\etc],'_', ID)
partial = re.sub('_+','_', partial) 
```

The second part is a bit more mysterious:

``` python
    # For safety, let's give it another scrub.
    result = []
    for word in _clean_re.split(partial):
        result.extend(unidecode(word).split())

    return unicode(''.join(result))
```

Seems like the processed partial is first split with the pattern defined in `re.compile` but that pattern contains characters like (back)slash which have been converted in the first part of the function, so why are we trying to match them again here?

Then each word is run through some weird `unidecode` and `split` operation, pushed onto an array, joined again and again munched by an `unicode` function => why? what is the purpose of these 4 lines of code?

**Comment by pieterdp** [comment]
@nvgeele Could you take a look at this?

**Comment by netsensei** [comment]
@nvgeele @Alina-Packed @pieterdp  _ping_

**Comment by pieterdp** [comment]
@netsensei : Nils is taking a look at it.

**Comment by nvgeele** [comment]
Indeed, the first part will iterate over the pattern/replacement pairs and apply them to the string. The replacements were taken from a document provided to me by Alina. The code may indeed be written a bit more efficient as you've already proposed. The exceptions still have to be done like they are now however.

The last part is done for removing all remaining illegal characters for which no replacement is/was specified (backslash is not filtered out by the first step as there is no replacement for it). Unidecode is used for, if I'm not mistaken, mapping unicode onto ASCII. All white-space should already be filtered out by the regex so the whole thing should be replaceable by

```
return ''.join([unidecode(w) for w in _clean_re.split(partial)])
```

There's probably a reason why I did it like that, but it's been a while since I wrote that code. If I find some time this weekend I'll make the changes and do the tests (currently doing research for my thesis).
